ktx2Decoder/src/wasmMemoryManager.ts

```diff
@@ -0,0 +1,82 @@
+declare function postMessage(message: any, transfer?: any[]): void;
+
+/**
+ * @hidden
+ */
+export class WASMMemoryManager {
+
+    public static LoadBinariesFromCurrentThread = true;
+    public static InitialMemoryPages = (1 * 1024 * 1024) >> 16; // 1 Mbytes
+
+    private static _RequestId = 0;
+
+    public static LoadWASM(path: string): Promise<ArrayBuffer> {
+        if (this.LoadBinariesFromCurrentThread) {
+            return new Promise((resolve, reject) => {
+                fetch(path)

```

Comment:
Browsers without Fetch API don't support WebAssembly, so maybe it's fine to use it here.

| API | Edge | Firefox | Chrome | Safari |
|-|-|-|-|-|
| Fetch | 14+ | 40+ | 42+ | 10.1+ (10.3+ on iOS) |
| WebAssembly | 16+ | 53+ | 57+ | 11+ |

IE11 supports neither.

Reply:
Given @lexaknyazev comment, let me know if I still should add custom code to download file.

Reply:
Hmm. IE11 certainly doesn't support WebAssembly, so maybe this specific part is okay. But this does raise a question for the individual transcoders. Should we have asmjs versions of them so that they can work in IE11? Draco, for example, has wasm and asmjs as fallback which makes it work with IE11. What do you think @lexaknyazev?

Reply:
IE11 supports neither BC7 nor ASTC. The only options for UASTC are BC1/BC3 or uncompressed RGBA32.

Reply:
Gotcha. What about individual transcoders for BC1/BC3 in the future? I think we would want to able use them in non-webassembly environments.

Reply:
I think adressing the problem of IE11 (if/when necessary) should be done in another PR?

Reply:
> I think adressing the problem of IE11 (if/when necessary) should be done in another PR?

Agree. Please open an issue in the transcoders repo if/when this gets relevant.

Reply:
Sounds good. We need to do this carefully as we still have legacy browser usage. In general, we can't use features not supported in IE11. This case is probably an exception.

Reply:
Here's a quick preview of what would it take to have an asm.js path based on couple tests. Note that this is based on UASTC/RGBA32 decoder.

- 120+ kB of minified JS (23 kB gzipped)
- slightly different usage
  ```js
  import * as decoder from './build/uastc_rgba32_unorm';

  // compute the memory footprint and allocate it
  decoder.memory.grow(texPages);

  // create compressed and decompressed memory views

  // copy compressed data

  // process
  decoder.decodeRGBA32(width, height);

  // copy decompressed data
  ```

Also note that transcoding UASTC textures with non-color data (such as ORM or normal maps) to anything except BC7 or ASTC would likely introduce noticeable artifacts, so RGBA32 target is still important for such cases.

Reply:
> Also note that transcoding UASTC textures with non-color data (such as ORM or normal maps) to anything except BC7 or ASTC would likely introduce noticeable artifacts, so RGBA32 target is still important for such cases.

Currently the transcoded format is determined automatically depending on the capability of the browser/GPU, it's not something the user can override => I don't know how it could work, notably in the GLTF loading assets context...

Reply:
To support the expected glTF workflow, a glTF loader should be able to request uncompressed target format when it needs higher quality. The most notable platform in this regard is macOS - its OpenGL driver (and therefore WebGL runtimes in Safari and Chrome) supports only BC1-BC5.

The same comment is also relevant when using UASTC textures with MSC transcoders.

Reply:
> a glTF loader should be able to request uncompressed target format when it needs higher quality

You saying depending on what the "type" of texture it is (i.e. ORM or normal maps vs color map), we should do different things right?

Reply:
Should the gltf loader really do this on its own, without a user request? Maybe in some cases the transcoding could be good enough even for non-color textures and the user would want to still save some space on the GPU?

Reply:
> You saying depending on what the "type" of texture it is (i.e. ORM or normal maps vs color map), we should do different things right?

Yes.

> Should the gltf loader really do this on its own, without a user request? Maybe in some cases the transcoding could be good enough even for non-color textures and the user would want to still save some space on the GPU?

3-channel normal maps will not work with BC1, ETC1, or PVRTC in almost all cases.